synapse/storage/databases/main/events.py

```diff
@@ -1504,6 +1504,9 @@ def _update_metadata_tables_txn(
 
             self._handle_event_relations(txn, event)
 
+            self._handle_insertion_event(txn, event)
+            self._handle_chunk_id(txn, event)
```

Comment:
Where is the best place to limit who can add an `insertion` event or a chunk connection content field?

Reply:
We probably want to reject such events, so we probably want to add some stuff to the event authorization code.

Reply:
I am having trouble fitting this into the existing code.

 - I need to make sure that only application services can send insertion, marker, and events with chunk ID's
 - Other homeservers need to be able to accept these events over federation. But probably should only trust the originating homeserver where the room is. How does originating homeserver even work when future room ID's don't include the homeserver?
    - Maybe it should be some power level where we can add the application service user ID?

---

 - [`synapse/event_auth.py` -> check(...)](https://github.com/matrix-org/synapse/blob/0ae95b38474a4d64a4d5057499e645a3b81e3736/synapse/event_auth.py):
    - It feels like the checks should go here next to `check_redaction`
    - That class doesn't have access to the `store` so I can't use `store.get_app_service_by_user_id(...)` to make sure the sender is an application service. Plus this code is probably run for events coming over federation which don't know which `sender` is an application service.
    - It could fit here if it was a new power level since we have the auth events
 - [`synapse/handlers/event_auth.py` -> `check_from_context(...)`](https://github.com/matrix-org/synapse/blob/0ae95b38474a4d64a4d5057499e645a3b81e3736/synapse/handlers/event_auth.py)
    - I can use `store.get_app_service_by_user_id(...)` but other code is still only protected by the raw `check(...)`

---

<details>
<summary>def _check_if_allowed_to_send_historical_events</summary>

```python
    def _check_if_allowed_to_send_historical_events(
        self,
        event: EventBase,
    ):
        """Check whether the event sender is allowed to add insertion, marker,
        or events with a chunk ID

        Raises:
            AuthError if the event sender is not an application service
        """

        if (
            event.type != EventTypes.MSC2716_INSERTION
            and event.type != EventTypes.MSC2716_MARKER
            and event.content.get(EventContentFields.MSC2716_CHUNK_ID) is None
        ):
            pass

        app_service = self._store.get_app_service_by_user_id(event.sender)

        if app_service is None:
            raise AuthError(403, "Only application services can send insertion events")
```

</details>

Reply:
Using the [` m.room.power_levels` -> `events`](https://matrix.org/docs/spec/client_server/r0.6.1#m-room-power-levels) ("The level required to send specific event types. This is a mapping from event type to power level required.") field seems perfect 🎉

Reply:
Added unstable room version `org.matrix.msc2716` which adds the `historical` power level that controls whether you can send `insertion`, `chunk`, and `marker` events ✅

I switched to `chunk` events so we can easily auth them against the PL level because they are sent by the application service user ID with the proper PL level for the room. See https://github.com/matrix-org/synapse/pull/10432